lib/ansible/module_utils/network/common/utils.py

```diff
@@ -442,6 +443,118 @@ def _fallback(fallback):
         pass
 
 
+def generate_dict(spec):
+    """
+    Generate dictionary which is in sync with argspec
+
+    :param spec: A dictionary which the argspec of module
+    :rtype: A dictionary
+    :returns: A dictionary in sync with argspec with default value
+    """
+    obj = {}
+    if not spec:
+        return obj
+
+    for key, val in iteritems(spec):
+        if 'default' in val:
+            dct = {key: val['default']}
+        elif 'type' in val and val['type'] == 'dict':
+            dct = {key: generate_dict(val['options'])}
+        else:
+            dct = {key: None}
+        obj.update(dct)
+    return obj
+
+
+def parse_conf_arg(cfg, arg):
+    """
+    Parse config based on argument
+
+    :param cfg: A text string which is a line of configuration.
+    :param arg: A text string which is to be matched.
+    :rtype: A text string
+    :returns: A text string if match is found
+    """
+    match = re.search(r'%s (.+)(\n|$)' % arg, cfg, re.M)
+    if match:
+        result = match.group(1).strip()
+    else:
+        result = None
+    return result
+
+
+def parse_conf_cmd_arg(cfg, cmd, res1, res2=None, delete_str='no'):
+    """
+    Parse config based on command
+
+    :param cfg: A text string which is a line of configuration.
+    :param cmd: A text string which is the command to be matched
+    :param res1: A text string to be returned if the command is present
+    :param res2: A text string to be returned if the negate command
+                 is present
+    :param delete_str: A text string to identify the start of the
+                 negate command
+    :rtype: A text string
+    :returns: A text string if match is found
+    """
+    match = re.search(r'\n\s+%s(\n|$)' % cmd, cfg)
+    if match:
+        return res1
+    if res2 is not None:
+        match = re.search(r'\n\s+%s %s(\n|$)' % (delete_str, cmd), cfg)
+        if match:
+            return res2
+    return None
+
+
+def get_xml_conf_arg(cfg, path, type='text'):
+    """
+    :param cfg: The top level configuration lxml Element tree object
+    :param path: The relative xpath w.r.t to top level element (cfg)
+           to be serached in the xml hierarchy
+    :param type: The type of data to be returned for the matched xml node.
+        Valid values are text, tag, attrib.
+    :return: Returns the required type for the matched xml node or else None
+    """
+    match = cfg.xpath(path)
+    if len(match):
+        result = getattr(match[0], type)
+    else:
+        result = None
+    return result
+
+
+def populate_config(cfg_dict):
```

Comment:
Can we rename this to something like remove_empties?

Reply:
Can we also add the ability to traverse a list of dicts? 

```    
    def generate_final_config(self, cfg_dict):
        """
        Generate final config dictionary

        :param cfg_dict: A dictionary parsed in the facts system
        :rtype: A dictionary
        :returns: A dictionary by eliminating keys that have null values
        """
        final_cfg = {}
        if not cfg_dict:
            return final_cfg

        for key, val in iteritems(cfg_dict):
            dct = None
            if isinstance(val, dict):
                child_val = self.generate_final_config(val)
                if child_val:
                    dct = {key: child_val}
            elif (isinstance(val, list) and val
                  and all([isinstance(x, dict) for x in val])):
                child_val = [self.generate_final_config(x) for x in val]
                if child_val:
                    dct = {key: child_val}
            elif val not in [None, [], {}, (), '']:
                dct = {key: val}
            if dct:
                final_cfg.update(dct)
        return final_cfg
``` 
something like this ^^^

Reply:
Fixed